ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py

```diff
@@ -0,0 +1,266 @@
+from typing import Any, Dict, List
+import numpy as np
+import tensorflow as tf
+
+from mlagents.envs.brain import BrainInfo
+from mlagents.trainers.buffer import Buffer
+from mlagents.trainers.components.reward_signals import RewardSignal, RewardSignalResult
+from mlagents.trainers.tf_policy import TFPolicy
+from .model import GAILModel
+from mlagents.trainers.demo_loader import demo_to_buffer
+
+
+class GAILRewardSignal(RewardSignal):
+    def __init__(
+        self,
+        policy: TFPolicy,
+        strength: float,
+        gamma: float,
+        demo_path: str,
```

Comment:
(discussed offline) I think it we could make this take a Buffer instance instead of the path, it would be a little cleaner. But might not be easy to pass it down from the TrainerController.

Reply:
One way we could do this is to add a `demo_path` config to the highest level (the PPO trainer), and if it exists, the TrainerController will load the demo and produce a demo Buffer, and pass it to the Trainer. Else it sets to None. Then the PreTraining and GAIL modules will attempt to grab the buffer from the Trainer and throw an exception if it is None. 

But this removes the ability to set a different demo file for each module separately and makes it a bit less compartmentalized from before - I'm not sure if at the end of the day it's overall cleaner